src/lib/locks.ts

```diff
@@ -0,0 +1,32 @@
+import {messageAllTabs, addListener, Message} from "@src/lib/messaging"
+
+export const OWNED_LOCKS = new Set()
+
+export async function acquire(lockname: string) {
+    if (OWNED_LOCKS.has(lockname)) return
+    async function getContention() {
+        return (await browser.runtime.sendMessage({type: "lock", command: "acquire", args: [lockname]})) ||
+        (await messageAllTabs("lock", "acquire", [lockname])).some(x => x === true)
```

Comment:
This doesn't seem to protect against the following scenario:
- Tab A messages all tabs asking if anybody has lock L
- Tab B messages all tabs asking if anybody has lock L
- Tab B answers that it doesn't have lock L to tab A
- Tab A answers that it doesn't have lock L to tab B
- Both tabs acquire lock L

I'm not sure how likely this scenario is to happen "in real life" but I have a feeling that it will happen from time to time.

Reply:
If A and B ask near-simultaneously enough, I think you're right. I suppose I need to read a textbook / wikipedia, e.g. https://www.wikipedia.org/wiki/Dekker%27s_algorithm

Reply:
https://www.geeksforgeeks.org/ricart-agrawala-algorithm-in-mutual-exclusion-in-distributed-system/ looks like a decent one to use. https://www.geeksforgeeks.org/mutual-exclusion-in-distributed-system/ has a nice summary.